eth-riscv-runtime/src/types.rs

```diff
@@ -12,12 +12,50 @@ use alloc::vec::Vec;
 #[derive(Default)]
 pub struct Mapping<K, V> {
     id: u64,
-    pd: PhantomData<(K, V)>
+    pd: PhantomData<(K, V)>,
+}
+
+/// A trait for types that can be read from and written to storage slots
+pub trait SlotStorable {
+    fn read_from_storage(key: u64) -> Self;
+    fn write_to_storage(&self, key: u64);
+}
+
+impl<V> SlotStorable for V
+where
+    V: SolValue + core::convert::From<<<V as SolValue>::SolType as SolType>::RustType>,
+{
+    fn read_from_storage(encoded_key: u64) -> Self {
+        let bytes: [u8; 32] = sload(encoded_key).to_be_bytes();
+        Self::abi_decode(&bytes, false).unwrap_or_else(|_| revert())
```

Comment:
do we want to use abi as storage encoding as fallback or in general? If we want to have Solidity compat I guess this works for all value types, but would be different for arrays and structs. But I guess we can write specific impls for those.